File: BinarySearch/Easy/BinarySearch.ts
// 704. Binary Search

function search(nums: number[], target: number): number {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    const midPoint = left + Math.floor((right - left) / 2);
    if (target == nums[midPoint]) {
      return midPoint;
    }

    if (target < nums[midPoint]) {
      right = midPoint - 1;
    }

    if (target > nums[midPoint]) {
      left = midPoint + 1;
    }
  }

  return -1;
}

// console.log(search([-1, 0, 3, 5, 9, 12], 9)); // 4
// console.log(search([5], 5)); // 0
console.log(search([-1, 0, 3, 5, 9, 12], 13)); // -1

File: BinarySearch/Easy/BinarySearch.py
# 704. Binary Search

from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        leftPointer = 0
        rightPointer = len(nums)

        while (leftPointer < rightPointer):
            midPointer = (rightPointer + leftPointer) // 2
            midPointerValue = nums[midPointer]
            if (target == midPointerValue):
                return midPointer

            if (target > midPointerValue):
                leftPointer = midPointer + 1
            else:
                rightPointer = midPointer

        return -1
